jepsen/src/jepsen/control.clj

```diff
@@ -360,6 +360,35 @@
 
 (def ssh "A remote that does things via clj-ssh." (SSHRemote. nil))
 
+(defn- undo-command-changes
+  "The `wrap-sudo` and `wrap-cd` functions adds stuff in front of commands which
+  makes docker crap itself, this function undoes that."
+  [cmd]
+  (-> cmd
+      (str/replace-first "sudo -S -u root" "")
+      (str/replace-first "cd /; " "")))
+
+(defrecord DockerRemote [nodes session]
+  Remote
+  (connect [this host]
+    (assoc this :session (str "jepsen-n" (inc (.indexOf nodes host)))))
+  (disconnect! [this]
+    (dissoc this :session))
+  (execute! [this action]
+    (let [cmd (str "docker exec " (:session this) " " (-> action :cmd undo-command-changes))]
+      (sh "sh" "-c" cmd)))
+  (upload! [this local-paths remote-path _rest]
+    (doseq [local-path local-paths]
+        (let [cmd (str "docker cp " local-path " " (:session this) ":" remote-path)]
+          (sh "sh" "-c" cmd))))
+  (download! [this remote-paths local-path rest]
+    (apply sh "docker" "cp" (str (:session this) ":" remote-paths) local-path rest)))
+
+(defn docker
+  "A remote that does things via `docker exec` and `docker cp`."
+  [nodes]
+  (->DockerRemote nodes nil))
+
```

Comment:
It looks like this protocol couples hardcoded assumptions in the Jepsen library to particular node choices in the docker config, and... I'm kinda wary about that. To be honest, I'm kinda wary of *any* attempts to make jepsen.control a multi-backend system; this stuff is complicated, involves a lot of baked-in assumptions, and attempts to add local command implementations instead of ssh has had a way of creating tests that nobody else can run. I just ripped a whole bunch of similar code *out* of the Mongo tests this week, because they'd become basically unrunnable in standard Jepsen environments. :-/

So... I'm not shooting this down entirely--I am willing to merge! But I do want to warn you that this is an area that's been historically difficult, and that you might be signing yourself up for a decade of handling support requests. I'm also concerned that you might be using this in everyday development work going forward, and as a result, future patches you (or others!) submit might *not* be tested with normal SSH remotes, and break in truly interesting and time-consuming ways. Are you planning on testing future patches/tests you submit to Jepsen in both SSH and Docker remotes?

Reply:
I think I see your concern.

>  I just ripped a whole bunch of similar code out of the Mongo tests this week

Is that change publicly available somewhere so that I can have a look?

> Are you planning on testing future patches/tests you submit to Jepsen in both SSH and Docker remotes?

If somebody make a change unrelated to `Remote`, what is it that can work with one instance of `Remote` but be broken with another? It seems to me that if you have an interface with two implementations that are correct, then you should be able to assume that everything that depend on those interfaces will work the same -- that's the whole point of an interface. Or am I missing something? (Assuming that the hardcoded assumptions are removed.)

Reply:
> It seems to me that if you have an interface with two implementations that are correct, then you should be able to assume that everything that depend on those interfaces will work the same -- that's the whole point of an interface.

I desperately want this to be true, and I think it might mostly be true here, which is why I'm willing to merge! All abstractions are leaky, though, and abstractions around things like "operating systems" or "running commands" tend to be particularly so--when tests have done things like this in the past, they tend to accrue increasing layers of environment-specific dispatch code to work around variances in the environments. MongoDB's test had hundreds of lines devoted to "local vs VM" environment differences. I'm hoping that will be less of an issue here.

For the actual changes, uh... see https://github.com/jepsen-io/mongodb. After n hours of trying to unwind the indirection layers they'd wrapped around jepsen.control, I actually deleted everything and started from scratch. :(

Reply:
tl;dr: I'm one person, working full time, as hard as I can, to make Jepsen a thing. But I can't test or support every combination of platforms, which means there's exactly two "blessed" configurations of Jepsen: [Debian 9] x [SSH remotes] x [iptables] x [LXC, real nodes]. Other implementations of the OS, Remote, Net, protocols, other environments... they're there so you *can* do this kind of polymorphism, but maintaining those implementations is your responsibility, and it can't break the blessed configurations. If you're OK with that, we'll merge (after sorting out the string munging stuff).

Reply:
> It looks like this protocol couples hardcoded assumptions in the Jepsen library to particular node choices in the docker config, and... I'm kinda wary about that.

Sorry for interrupting. FWIW a while ago I made an attempt to overcome this through asking docker daemon for container id with a specific label usually assigned by docker-compose during setup phase.

https://github.com/rbkmoney/riak-jepsenized/blob/fa3cf723dd494da1becb52f9e1d6897d071ca27a/src/jepsenized/riak/control/docker.clj#L27-L30

In retrospect I think I had to name it `DockerComposeRemote` though.

Reply:
> I just ripped a whole bunch of similar code out of the Mongo tests this week

> For the actual changes, uh... see https://github.com/jepsen-io/mongodb

That repo hasn't changed in almost two years?

Reply:
Haven't pushed yet.

Reply:
@keynslug Cool, I'll have a look at your solution in detail. Thanks for making me aware of it!